src/utils.rs

```diff
@@ -0,0 +1,439 @@
+//! }
+//!
+//! let mut rng = Step4x64Rng::seed_from_u64(42);
+//!
+//! assert_eq!(rng.next_u32(), 0x7ba1_8fa4);
+//! assert_eq!(rng.next_u64(), 0xb814_0169_cca1_b8ea);
+//! let mut buf = [0u8; 5];
+//! rng.fill_bytes(&mut buf);
+//! assert_eq!(buf, [0x2b, 0x8c, 0xc8, 0x75, 0x18]);
+//! ```
+//!
+//! ## Fill-based RNG
+//!
+//! ```
+//! use rand_core::RngCore;
+//!
+//! pub struct FillRng(u8);
+//!
+//! impl RngCore for FillRng {
+//!     fn next_u32(&mut self) -> u32 {
+//!         let mut buf = [0; 4];
+//!         self.fill_bytes(&mut buf);
+//!         u32::from_le_bytes(buf)
+//!     }
+//!
+//!     fn next_u64(&mut self) -> u64 {
+//!         let mut buf = [0; 8];
+//!         self.fill_bytes(&mut buf);
+//!         u64::from_le_bytes(buf)
+//!     }
+//!
+//!     fn fill_bytes(&mut self, dst: &mut [u8]) {
+//!         for byte in dst {
+//!             let val = self.0;
+//!             self.0 = val + 1;
+//!             *byte = val;
+//!         }
+//!     }
+//! }
+//!
+//! let mut rng = FillRng(0);
+//!
+//! assert_eq!(rng.next_u32(), 0x03_020100);
+//! assert_eq!(rng.next_u64(), 0x0b0a_0908_0706_0504);
+//! let mut buf = [0u8; 5];
+//! rng.fill_bytes(&mut buf);
+//! assert_eq!(buf, [0x0c, 0x0d, 0x0e, 0x0f, 0x10]);
+//! ```
+//!
+//! Note that you can use `from_ne_bytes` instead of `from_le_bytes`
+//! if your `fill_bytes` implementation is not reproducible.
+
+use crate::RngCore;
+
+/// Implement `next_u64` via `next_u32` using little-endian order.
+pub fn next_u64_via_u32<R: RngCore + ?Sized>(rng: &mut R) -> u64 {
+    // Use LE; we explicitly generate one value before the next.
+    let x = u64::from(rng.next_u32());
+    let y = u64::from(rng.next_u32());
+    (y << 32) | x
+}
+
+/// Implement `fill_bytes` via `next_u64` using little-endian order.
+pub fn fill_bytes_via_next_word<W: Word>(dest: &mut [u8], mut next_word: impl FnMut() -> W) {
+    let mut chunks = dest.chunks_exact_mut(size_of::<W>());
+    for chunk in &mut chunks {
+        let val = next_word();
+        chunk.copy_from_slice(val.to_le_bytes().as_ref());
+    }
+    let rem = chunks.into_remainder();
+    if !rem.is_empty() {
+        let val = next_word().to_le_bytes();
+        rem.copy_from_slice(&val.as_ref()[..rem.len()]);
+    }
+}
+
+/// Fills slice of words `dst` from byte slice `src` using little endian order.
+///
+/// # Panics
+///
+/// If `size_of_val(src) != size_of_val(dst)`.
+#[inline]
+pub fn read_words_into<W: Word>(src: &[u8], dst: &mut [W]) {
+    assert_eq!(size_of_val(src), size_of_val(dst));
+    let chunks = src.chunks_exact(size_of::<W>());
+    for (out, chunk) in dst.iter_mut().zip(chunks) {
+        let Ok(bytes) = chunk.try_into() else {
+            unreachable!()
+        };
+        *out = W::from_le_bytes(bytes);
+    }
+}
+
+/// Create new block buffer.
+///
+/// # Panics
+/// If `N` is smaller than 2 or bigger than 2^32 - 1.
+pub fn new_buffer<W: Word, const N: usize>() -> [W; N] {
+    assert!(N > 2);
+    assert!(N < 1 << 32);
```

Comment:
`1 << 32` will overflow on 32-bit platforms. Maybe use `N <= u32::MAX as usize`?

Reply:
I replaced the assert with `W::from_usize(N)`, which is equivalent to assert that `N` is not bigger than `W::MAX`.